pkg/multicluster/cluster_management.go

```diff
@@ -17,174 +17,416 @@ limitations under the License.
+	csrCreationTracker.FinalMSG = "Successfully found corresponding CSR from the agent.\n"
+	csrCreationTracker.Start()
+	if err := hubCluster.WaitForCSRCreated(ctx, spokeCluster.Name); err != nil {
+		return errors.Wrap(err, "failed found CSR created by registration agent")
+	}
+	csrCreationTracker.Stop()
+
+	args.ioStreams.Infof("Approving the CSR for cluster %q.\n", spokeCluster.Name)
+	if err := hubCluster.ApproveCSR(ctx, spokeCluster.Name); err != nil {
+		return errors.Wrap(err, "failed found CSR created by registration agent")
+	}
+
+	ready, err := hubCluster.WaitForSpokeClusterReady(ctx, clusterConfig.ClusterName)
+	if err != nil || !ready {
+		return errors.Errorf("fail to waiting for register request")
+	}
+
+	if err = hubCluster.RegisterSpokeCluster(ctx, spokeCluster.Name); err != nil {
+		return errors.Wrap(err, "fail to approve spoke cluster")
+	}
+	return nil
 }
 
-// JoinClusterByKubeConfig add child cluster by kubeconfig path, return cluster info and error
-func JoinClusterByKubeConfig(_ctx context.Context, k8sClient client.Client, kubeconfigPath string, clusterName string) (*api.Cluster, error) {
-	config, err := clientcmd.LoadFromFile(kubeconfigPath)
+// LoadKubeClusterConfigFromFile create KubeClusterConfig from kubeconfig file
+func LoadKubeClusterConfigFromFile(filepath string) (*KubeClusterConfig, error) {
+	clusterConfig := &KubeClusterConfig{}
+	var err error
+	clusterConfig.Config, err = clientcmd.LoadFromFile(filepath)
 	if err != nil {
 		return nil, errors.Wrapf(err, "failed to get kubeconfig")
 	}
-	if len(config.CurrentContext) == 0 {
+	if len(clusterConfig.Config.CurrentContext) == 0 {
 		return nil, fmt.Errorf("current-context is not set")
 	}
-	ctx, ok := config.Contexts[config.CurrentContext]
+	var ok bool
+	ctx, ok := clusterConfig.Config.Contexts[clusterConfig.Config.CurrentContext]
 	if !ok {
-		return nil, fmt.Errorf("current-context %s not found", config.CurrentContext)
+		return nil, fmt.Errorf("current-context %s not found", clusterConfig.Config.CurrentContext)
 	}
-	cluster, ok := config.Clusters[ctx.Cluster]
+	clusterConfig.Cluster, ok = clusterConfig.Config.Clusters[ctx.Cluster]
 	if !ok {
 		return nil, fmt.Errorf("cluster %s not found", ctx.Cluster)
 	}
-	authInfo, ok := config.AuthInfos[ctx.AuthInfo]
+	clusterConfig.AuthInfo, ok = clusterConfig.Config.AuthInfos[ctx.AuthInfo]
 	if !ok {
 		return nil, fmt.Errorf("authInfo %s not found", ctx.AuthInfo)
 	}
+	clusterConfig.ClusterName = ctx.Cluster
+	if endpoint, err := utils.ParseAPIServerEndpoint(clusterConfig.Cluster.Server); err == nil {
+		clusterConfig.Cluster.Server = endpoint
+	} else {
+		_, _ = fmt.Fprintf(&clusterConfig.Logs, "failed to parse server endpoint: %v", err)
+	}
+	return clusterConfig, nil
+}
+
+const (
+	// ClusterGateWayEngine cluster-gateway cluster management solution
+	ClusterGateWayEngine = "cluster-gateway"
+	// OCMEngine ocm cluster management solution
+	OCMEngine = "ocm"
```

Comment:
what's the difference between these two engines? AFAIK, using OCM we also need cluster-gateway.

Reply:
For `cluster-gateway` (by default), the implementation of `cluster` is a secret. For `ocm`, the implementation of `cluster` is OCM ManagedCluster. The join and detach logic will differ a lot. The name of `engine` here is a legacy design by @yue9944882 . Normally, we do not need to care about them, except when users want to use OCM.

Reply:
OK